prow/pod-utils/decorate/podspec.go

```diff
@@ -0,0 +1,311 @@
+
+	"k8s.io/test-infra/prow/kube"
+)
+
+const (
+	JobNameEnv   = "JOB_NAME"
+	JobSpecEnv   = "JOB_SPEC"
+	JobTypeEnv   = "JOB_TYPE"
+	ProwJobIdEnv = "PROW_JOB_ID"
+
+	BuildIdEnv = "BUILD_ID"
+	// Deprecated, will be removed in the future.
+	ProwBuildIdEnv = "BUILD_NUMBER"
+	// Deprecated, will be removed in the future.
+	JenkinsBuildIdEnv = "buildId"
+
+	RepoOwnerEnv   = "REPO_OWNER"
+	RepoNameEnv    = "REPO_NAME"
+	PullBaseRefEnv = "PULL_BASE_REF"
+	PullBaseShaEnv = "PULL_BASE_SHA"
+	PullRefsEnv    = "PULL_REFS"
+	PullNumberEnv  = "PULL_NUMBER"
+	PullPullShaEnv = "PULL_PULL_SHA"
+)
+
+// EnvForSpec returns a mapping of environment variables
+// to their values that should be available for a job spec
+func EnvForSpec(spec JobSpec) (map[string]string, error) {
+	env := map[string]string{
+		JobNameEnv:   spec.Job,
+		BuildIdEnv:   spec.BuildId,
+		ProwJobIdEnv: spec.ProwJobId,
+		JobTypeEnv:   string(spec.Type),
+	}
+
+	// for backwards compatibility, we provide the build ID
+	// in both $BUILD_ID and $BUILD_NUMBER for Prow agents
+	// and in both $buildId and $BUILD_NUMBER for Jenkins
+	if spec.agent == kube.KubernetesAgent {
+		env[ProwBuildIdEnv] = spec.BuildId
+	} else if spec.agent == kube.JenkinsAgent {
+		env[JenkinsBuildIdEnv] = spec.BuildId
+	}
+
+	raw, err := json.Marshal(spec)
+	if err != nil {
+		return env, fmt.Errorf("failed to marshal job spec: %v", err)
+	}
+	env[JobSpecEnv] = string(raw)
+
+	if spec.Type == kube.PeriodicJob {
+		return env, nil
+	}
+	env[RepoOwnerEnv] = spec.Refs.Org
+	env[RepoNameEnv] = spec.Refs.Repo
+	env[PullBaseRefEnv] = spec.Refs.BaseRef
+	env[PullBaseShaEnv] = spec.Refs.BaseSHA
+	env[PullRefsEnv] = spec.Refs.String()
+
+	if spec.Type == kube.PostsubmitJob || spec.Type == kube.BatchJob {
+		return env, nil
+	}
+	env[PullNumberEnv] = strconv.Itoa(spec.Refs.Pulls[0].Number)
+	env[PullPullShaEnv] = spec.Refs.Pulls[0].SHA
+	return env, nil
+}
+
+func EnvForType(jobType kube.ProwJobType) []string {
+	baseEnv := []string{JobNameEnv, JobSpecEnv, JobTypeEnv, ProwJobIdEnv, BuildIdEnv, ProwBuildIdEnv, JenkinsBuildIdEnv}
+	refsEnv := []string{RepoOwnerEnv, RepoNameEnv, PullBaseRefEnv, PullBaseShaEnv, PullRefsEnv}
+	pullEnv := []string{PullNumberEnv, PullPullShaEnv}
+
+	switch jobType {
+	case kube.PeriodicJob:
+		return baseEnv
+	case kube.PostsubmitJob, kube.BatchJob:
+		return append(baseEnv, refsEnv...)
+	case kube.PresubmitJob:
+		return append(append(baseEnv, refsEnv...), pullEnv...)
+	default:
+		return []string{}
+	}
+}
+
+func Labels() []string {
+	return []string{kube.ProwJobTypeLabel, kube.CreatedByProw}
+}
+
+func VolumeMounts() []string {
+	return []string{LogMountName, CodeMountName, ToolsMountName, GcsCredentialsMountName}
+}
+
+func VolumeMountPaths() []string {
+	return []string{LogMountPath, CodeMountPath, ToolsMountPath, GcsCredentialsMountPath}
+}
+
+const (
+	LogMountName            = "logs"
+	LogMountPath            = "/logs"
+	CodeMountName           = "code"
+	CodeMountPath           = "/go"
```

Comment:
Why go?

Reply:
/shrug

Reply:
I would suggest something like `/home/go/`, anything with *at least* one layer deeper of nesting. Preferably it would actually be 3 deep (`/home/prow/go` ?). Why?
- we really shouldn't make `$HOME=/`, that's going to be "fun"
- we really ought to put code under a `$HOME` so various tools behave well
- we really ought to be able to use alias support to check out some code to another `GOPATH` within home or even another `HOME`.